go/store/types/edits/async_sorted_edits.go

```diff
@@ -184,87 +112,42 @@ func (ase *AsyncSortedEdits) Close() {
 	}
 }
 
-func (ase *AsyncSortedEdits) wait() {
-	running := ase.asyncConcurrency
-
-	for running > 0 {
-		select {
-		case val := <-ase.resultChan:
-			coll := NewKVPCollection(ase.nbf, val)
-			ase.sortedColls = append(ase.sortedColls, coll)
-
-		case <-ase.doneChan:
-			running--
-		}
-	}
-
-	for {
-		select {
-		case val := <-ase.resultChan:
-			coll := NewKVPCollection(ase.nbf, val)
-			ase.sortedColls = append(ase.sortedColls, coll)
-		default:
-			close(ase.resultChan)
-			return
-		}
-	}
-}
-
-// Sort performs a concurrent merge sort.  Once this completes use the Iterator method for getting a KVPIterator
-// which can be used to iterate over all the KVPs in order.
-func (ase *AsyncSortedEdits) Sort() {
+// mergeCollections performs a concurrent sorted-merge of |sortedColls|. Must be called after |sortGroup| is complete.
+// Once this completes use the |iterator| method for getting a KVPIterator which can be used to iterate over all the
+// KVPs in order.
+func (ase *AsyncSortedEdits) mergeCollections() error {
+	sema := semaphore.NewWeighted(int64(ase.sortConcurrency))
 	for len(ase.sortedColls) > 2 {
 		pairs := pairCollections(ase.sortedColls)
-		ase.sortedColls = nil
-
-		numPairs := len(pairs)
-
-		numGoRs := ase.sortConcurrency
-		if numGoRs > numPairs {
-			numGoRs = numPairs
-		}
-
-		sortChan := make(chan [2]*KVPCollection, numPairs)
-		resChan := make(chan *KVPCollection, numPairs)
-		for i := 0; i < numGoRs; i++ {
-			go func() {
-				defer func() {
-					ase.doneChan <- true
-				}()
-
-				err := merger(sortChan, resChan)
-				ase.ae.SetIfError(err)
-			}()
-		}
-
-		for _, pair := range pairs {
-			sortChan <- pair
-		}
-
-		close(sortChan)
-
-		for numGoRs > 0 {
-			select {
-			case val := <-resChan:
-				ase.sortedColls = append(ase.sortedColls, val)
-
-			case <-ase.doneChan:
-				numGoRs--
+		ase.sortedColls = make([]*KVPCollection, len(pairs))
```

Comment:
Once we have our pairs, we know we're going to have one result collection per pair, and they go can directly into `sortedColls`.